File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/go/pkg/bundle/bundle.go
// Package bundle provides trust bundles for offline/air-gapped SchemaPin verification.
package bundle

import (
	"encoding/json"
	"fmt"

	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
	"github.com/ThirdKeyAi/schemapin/go/pkg/revocation"
)

// BundledDiscovery combines a domain with its well-known response.
// Uses custom JSON marshaling for flattened format.
type BundledDiscovery struct {
	Domain   string
	WellKnown discovery.WellKnownResponse
}

// MarshalJSON implements custom JSON marshaling with flattened format.
func (b BundledDiscovery) MarshalJSON() ([]byte, error) {
	m := map[string]interface{}{
		"domain":         b.Domain,
		"schema_version": b.WellKnown.SchemaVersion,
		"developer_name": b.WellKnown.DeveloperName,
		"public_key_pem": b.WellKnown.PublicKeyPEM,
	}
	if b.WellKnown.Contact != "" {
		m["contact"] = b.WellKnown.Contact
	}
	if len(b.WellKnown.RevokedKeys) > 0 {
		m["revoked_keys"] = b.WellKnown.RevokedKeys
	}
	if b.WellKnown.RevocationEndpoint != "" {
		m["revocation_endpoint"] = b.WellKnown.RevocationEndpoint
	}
	return json.Marshal(m)
}

// UnmarshalJSON implements custom JSON unmarshaling from flattened format.
func (b *BundledDiscovery) UnmarshalJSON(data []byte) error {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return err
	}

	if v, ok := m["domain"]; ok {
		if err := json.Unmarshal(v, &b.Domain); err != nil {
			return fmt.Errorf("failed to unmarshal domain: %w", err)
		}
	}

	// Unmarshal the well-known fields directly from the flat map
	wellKnownData, err := json.Marshal(m)
	if err != nil {
		return err
	}
	return json.Unmarshal(wellKnownData, &b.WellKnown)
}

// SchemaPinTrustBundle holds discovery documents and revocations for offline use.
type SchemaPinTrustBundle struct {
	SchemapinBundleVersion string                       `json:"schemapin_bundle_version"`
	CreatedAt              string                       `json:"created_at"`
	Documents              []BundledDiscovery           `json:"documents"`
	Revocations            []revocation.RevocationDocument `json:"revocations"`
}

// NewTrustBundle creates a new empty trust bundle.
func NewTrustBundle(createdAt string) *SchemaPinTrustBundle {
	return &SchemaPinTrustBundle{
		SchemapinBundleVersion: "1.2",
		CreatedAt:              createdAt,
		Documents:              []BundledDiscovery{},
		Revocations:            []revocation.RevocationDocument{},
	}
}

// FindDiscovery finds a discovery document for a domain.
func (b *SchemaPinTrustBundle) FindDiscovery(domain string) *discovery.WellKnownResponse {
	for i := range b.Documents {
		if b.Documents[i].Domain == domain {
			return &b.Documents[i].WellKnown
		}
	}
	return nil
}

// FindRevocation finds a revocation document for a domain.
func (b *SchemaPinTrustBundle) FindRevocation(domain string) *revocation.RevocationDocument {
	for i := range b.Revocations {
		if b.Revocations[i].Domain == domain {
			return &b.Revocations[i]
		}
	}
	return nil
}

// ParseTrustBundle parses a trust bundle from a JSON string.
func ParseTrustBundle(jsonStr string) (*SchemaPinTrustBundle, error) {
	var bundle SchemaPinTrustBundle
	if err := json.Unmarshal([]byte(jsonStr), &bundle); err != nil {
		return nil, fmt.Errorf("failed to parse trust bundle: %w", err)
	}
	return &bundle, nil
}

File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/rust/src/types/bundle.rs
use serde::{Deserialize, Serialize};

use super::discovery::WellKnownResponse;
use super::revocation::RevocationDocument;

/// A pre-shared collection of discovery and revocation documents for offline verification.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SchemaPinTrustBundle {
    pub schemapin_bundle_version: String,
    pub created_at: String,
    pub documents: Vec<BundledDiscovery>,
    #[serde(default)]
    pub revocations: Vec<RevocationDocument>,
}

/// A discovery document bundled with its domain.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct BundledDiscovery {
    pub domain: String,
    #[serde(flatten)]
    pub well_known: WellKnownResponse,
}

impl SchemaPinTrustBundle {
    /// Create a new empty trust bundle.
    pub fn new(created_at: &str) -> Self {
        Self {
            schemapin_bundle_version: "1.2".to_string(),
            created_at: created_at.to_string(),
            documents: vec![],
            revocations: vec![],
        }
    }

    /// Find a discovery document by domain.
    pub fn find_discovery(&self, domain: &str) -> Option<&BundledDiscovery> {
        self.documents.iter().find(|d| d.domain == domain)
    }

    /// Find a revocation document by domain.
    pub fn find_revocation(&self, domain: &str) -> Option<&RevocationDocument> {
        self.revocations.iter().find(|r| r.domain == domain)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_test_bundle() -> SchemaPinTrustBundle {
        SchemaPinTrustBundle {
            schemapin_bundle_version: "1.2".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![BundledDiscovery {
                domain: "example.com".to_string(),
                well_known: WellKnownResponse {
                    schema_version: "1.2".to_string(),
                    developer_name: Some("Example Corp".to_string()),
                    public_key_pem: "-----BEGIN PUBLIC KEY-----\ntest\n-----END PUBLIC KEY-----"
                        .to_string(),
                    revoked_keys: vec![],
                    contact: None,
                    revocation_endpoint: None,
                },
            }],
            revocations: vec![],
        }
    }

    #[test]
    fn test_trust_bundle_serde_roundtrip() {
        let bundle = make_test_bundle();
        let json = serde_json::to_string_pretty(&bundle).unwrap();
        let bundle2: SchemaPinTrustBundle = serde_json::from_str(&json).unwrap();
        assert_eq!(bundle, bundle2);
    }

    #[test]
    fn test_find_discovery() {
        let bundle = make_test_bundle();
        assert!(bundle.find_discovery("example.com").is_some());
        assert!(bundle.find_discovery("other.com").is_none());
    }

    #[test]
    fn test_find_revocation() {
        let bundle = make_test_bundle();
        assert!(bundle.find_revocation("example.com").is_none());
    }

    #[test]
    fn test_new_bundle() {
        let bundle = SchemaPinTrustBundle::new("2026-02-10T00:00:00Z");
        assert_eq!(bundle.schemapin_bundle_version, "1.2");
        assert!(bundle.documents.is_empty());
        assert!(bundle.revocations.is_empty());
    }

    #[test]
    fn test_bundled_discovery_flattening() {
        let bd = BundledDiscovery {
            domain: "example.com".to_string(),
            well_known: WellKnownResponse {
                schema_version: "1.2".to_string(),
                developer_name: Some("Test".to_string()),
                public_key_pem: "key".to_string(),
                revoked_keys: vec![],
                contact: None,
                revocation_endpoint: None,
            },
        };
        let json = serde_json::to_string(&bd).unwrap();
        // Flattened fields should appear at the top level
        assert!(json.contains("\"domain\""));
        assert!(json.contains("\"schema_version\""));
        assert!(json.contains("\"public_key_pem\""));
    }
}
